### ISSUE TITLE
Future of `Let` vernacular

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#10459]
As discussed in #10446 , the current implementation of `Let` is inconsistent, for example opacity is not always respected, and unfolding can make such definitions disappear which looks quite fragile.

Personally, I'm not sure about the benefits of `Let` so I would maybe just deprecate it altogether, but I'm sure users won't like this.

Nevertheless, something has to be done to bring consistency to the implementation.

We have are several alternatives:
- Deprecate `Let` in its current form, and replace it by non-clearable section variables.
- Deprecated opaque `Let` . Indeed inconsistencies regarding opaque `Let` is what initiated the discussion so this would make sense. Doing this was done in #10446 and indeed it broke a couple of developments, but IMHO with good reason.
- Make behavior of `Let` consistent; we can profit from some recent refactorings and have all paths  respect "opacity".

WDYT folks?

**Comment by SkySkimmer** [comment]
Opaque Let has nonsense semantics imo
Transparent Let is fine AFAICT
>Deprecate Let in its current form, and replace it by non-clearable section variables.

I don't know what this means.

**Comment by ejgallego** [comment]
> Opaque Let has nonsense semantics imo

Wouldn't it be the same that a section variable, but that doesn't need discharging? 

> I don't know what this means.

I mean why do we have let and section variables, aren't they the same?

I also mean, isn't it very fragile to allow users to clear section variables [I know, never seen topic]

**Comment by SkySkimmer** [comment]
>Wouldn't it be the same that a section variable, but that doesn't need discharging?

What is this supposed to say?
~~~coq
Section foo.
Let bla : nat.
Proof. exact 0. Qed.
Definition bli := bla.
End foo.
Compute bli. (* 0 *)
~~~

>I mean why do we have let and section variables, aren't they the same?

Is `fun x =>` and `let x := bla in` the same?

**Comment by ejgallego** [comment]
> What is this supposed to say?

I would expect it to behave like `Definition` [I'm likely missing something]



> Is `fun x =>` and `let x := bla in` the same?

The thing is that we don't have lambda at the vernacular level, do we?

**Comment by SkySkimmer** [comment]
>The thing is that we don't have lambda at the vernacular level, do we?

It's called `Variable`

**Comment by ejgallego** [comment]
> It's called `Variable`

Yup then we fall into the domain of #6254 , I guess.

**Comment by gares** [comment]
FTR, Let is the only way to define something is a section and have it disappear for real from the global namespace at closing time, so it has a use case.

 I prefer it was implemented with a constant and an inline attribute, rather than a context entry (and that would fix the opacity discrepancy).

**Comment by ppedrot** [comment]
I stand with @SkySkimmer on this one: Let is definitely useful (and needed), but Qed-terminated Let doesn't make any sense.

**Comment by ggonthier** [comment]
> Qed-terminated Let doesn't make any sense.
I have to disagree with this. A `Qed` terminated `Let` is the natural way of stating an internal technical lemma, so that it is not exported (even as a qualified name) from its section or file, and is automatically added as a `Hint`. See the Odd Order theorem proof for numerous instances of this use.

**Comment by ppedrot** [comment]
@ggonthier To be fair, I am not surprised that there are reasonable uses of Qed-terminated lets, but the use you mention looks more like an XY-problem to me. You could get a similar behaviour with Defined-terminated lets + clearbody at proof opening (which is essentially what the implementation performs). My remark was more applying to the "kernel" semantics of Qed lets (as I am trying to move part of the section mechanism into the kernel, where it belongs TCB-wise).

**Comment by ppedrot** [comment]
By the way, the implementation has interesting bugs:
```
Section Foo.

Let b : bool.
Proof.
exact true.
Qed.

Definition foo :=
match b return if b then unit else bool with
| true => tt
| false => false
end.

Goal foo = tt.
Proof.
constructor. (* goal is actually ill-typed *)
```

I didn't try to break it too much, but it's probably possible to trigger fancy anomalies with that.
